catalog/newcontext/products.py

```diff
@@ -108,5 +112,35 @@ def context(self):
         }
 
         return {
+            **self._products.context(),
             'product_images': product_images,
         }
+
+
+class PaginatedProducts(Products):
+    """Slice products and add pagination data to a context."""
+
+    def __init__(self, products: Products, url: str, page_number: int, per_page: int):
+        if (
+            page_number < 1 or
+            per_page not in settings.CATEGORY_STEP_MULTIPLIERS
+        ):
+            raise http.Http404('Page does not exist.')
+
+        self._products = products
+        self._url = url
+        self._page_number = page_number
+        self._pagination = PaginationContext(url, page_number, per_page, self._products.qs())
+
+    @lru_cache()
+    def _pagination_context(self):
+        return self._pagination.context()
+
+    def qs(self):
+        return self._pagination_context()['page'].object_list
+
+    def context(self):
+        return {
+            **self._products.context(),
```

Comment:
It just helps to avoid key conflicts, that would be occurred. For example we already have got `page` instance in final context, but pagination has `page` too